### ISSUE TITLE
Add support for ALiBi/relative positional biases to the fast path for Transformers

### ISSUE THREAD

**Issue by ani300** [pytorch/pytorch#96099]
### 🚀 The feature, motivation and pitch

Recent work in transformer architectures such as T5 [1] and Attention with Linear Biases (ALiBi) [2] have shown that moving the positional encodings from the word embedding layer and directly into the self-attention computation improves the capacity of models to extrapolate to longer sentences while keeping lower perplexity scores. Both T5 and ALiBi add biases to the logits resulting from the QK multiplication, turning the softmax operation into something like this:

weights = softmax(QK^T + B), where B can either be a constant pre-defined matrix, like in ALiBi, or a learnable bias, like in T5. In both cases, the values of B are dependent on the relative distance between tokens, with the farthest tokens usually having the highest penalties.

Usually, this is implemented through modification of the attention mask to add these extra weights on top, changing it from a boolean yes/no attention to a weighted kind of model.

Performance-wise, our experiments are summarized in the graph below, with @daviswer explaining: "Here we train 3 AliBi decoders (390M params) with different sequence lengths (2048, 1024, 512) but perform validation at sequence length 2048 all the time. This allows us to track the degree of overfitting to sequence length over time. We then do the same for 3 equivalent models using T5 relative positional bias (rpb) instead of AliBi. While val perplexity of both variants suffers greatly over time with diverging train/val sequence lengths, the degree of increase is worse for rpb, indicating that while generalization to longer sequence lengths is a difficult problem, AliBi is indeed better at it than rpb." From the ALiBi paper we also know that both perform better than the traditional word embedding-level positional encodings.

![loss_curve (1)](https://user-images.githubusercontent.com/919977/223146953-d460bf47-29fe-482f-8f85-13376c686bab.png)

As of right now, the BetterTransformers path in PyTorch only supports two models for the attention mask: none at all, or the one used for causal LMs. We propose modifying the current implementation so that more complex masks like the ones required for relative positional biases and ALiBi can also be supported.

cc: @daviswer @supriyogit @raghukiran1224 @mudhakar @mayank31398 @cpuhrsch @HamidShojanazeri

[1] Exploring the Limits of Transfer Learning with a Unified Text-to-Text Transformer (https://arxiv.org/abs/1910.10683)

[2] Train Short, Test Long: Attention with Linear Biases Enables Input Length Extrapolation (https://arxiv.org/abs/2108.12409)

cc @jbschlosser @bhosmer @cpuhrsch @erichan1

**Comment by cpuhrsch** [comment]
cc @drisspg

**Comment by drisspg** [comment]
We are planning on upstreaming the changes introduced in  Xformers for their mem-efficient kernel that allow for an arbitrary mask to be fused into SDPA. Causal is somewhat special since the mask has a well defined pattern and doesn't require the kernel to materialize it. 

It would be kind of nice to create a mask "zoo" which could provide a nice abstraction layer the different attention mask familys. I think that some more thought would need to be taken for this but once the above is introduced some perf gains will still be had for the fused kernels with arbitrary additive attention masks.

**Comment by ani300** [comment]
Do you have an approximate timeline on when that is going to happen?

**Comment by drisspg** [comment]
> Do you have an approximate timeline on when that is going to happen?

Likely within a couple of weeks

**Comment by chenjoya** [comment]
Any progress? It seems ```scaled_dot_product_attention``` produces much more memory and time costs with relative pos (attn_mask) provided. Thank u.

**Comment by drisspg** [comment]
> Any progress? It seems `scaled_dot_product_attention` produces much more memory and time costs with relative pos (attn_mask) provided. Thank u.

No update yet but will post back here after the upstream

**Comment by netw0rkf10w** [comment]
Hello,

According to the documentation, `F.scaled_dot_product_attention(query, key, value, attn_mask=None, dropout_p=0.0, is_causal=False, scale=None)` is equivalent to:

```python
# Efficient implementation equivalent to the following:
scale_factor = 1 / math.sqrt(Q.size(-1)) if scale is None else scale
attn_mask = torch.ones(L, S, dtype=torch.bool).tril(diagonal=0) if is_causal else attn_mask
attn_mask = attn_mask.masked_fill(not attn_mask, -float('inf')) if attn_mask.dtype==torch.bool else attn_mask
attn_weight = torch.softmax((Q @ K.transpose(-2, -1) * scale_factor) + attn_mask, dim=-1)
attn_weight = torch.dropout(attn_weight, dropout_p)
return attn_weight @ V
```

Thus, it suffices to feed into the model the bias tensor `B` by setting `attn_mask=B` to achieve what is requested in this issue (the only condition is that `B` is not a boolean tensor). 

Am I missing something? Thanks for anh clarifications.

**Comment by ani300** [comment]
This is correct, but doing so will send the execution of the operator through a slow path instead of a fast one, nullifying any performance gains we could get by using this instead of our operator in the first place. This issue is about supporting arbitrary masks for the fast execution paths.

**Comment by joecummings** [comment]
> > Any progress? It seems `scaled_dot_product_attention` produces much more memory and time costs with relative pos (attn_mask) provided. Thank u.
> 
> No update yet but will post back here after the upstream

@drisspg Is this confirmed to be going into the 2.0.1 release (that is happening soon)? Or is the timeline longer?

**Comment by drisspg** [comment]
No this won't be going into the 2.0.1 release since that is for bug fixes and this is a new feature, not a bug fix. We are planning for the 2.1.0 release.

**Comment by drisspg** [comment]
Just a quick updated here, we have been communicating with xformers and plan on upstreaming by end of april.

**Comment by ofirpress** [comment]
Hi @drisspg!

Am I correct in thinking that this discussion references enabling custom masks just for the torch.backends.cuda.enable_mem_efficient_sdp()
backend and not for the flash attention backend?

Are there any plans to enable inputting custom masks in flash attention?

Thanks!

**Comment by drisspg** [comment]
> Hi @drisspg!
> 
> 
> 
> Am I correct in thinking that this discussion references enabling custom masks just for the torch.backends.cuda.enable_mem_efficient_sdp()
> 
> backend and not for the flash attention backend?
> 
> 
> 
> Are there any plans to enable inputting custom masks in flash attention?
> 
> 
> 
> Thanks!

Correct this discussion is only around mem efficient attention. Currently FlashAttention does not support attention bias but you can see that it is on their roadmap posted on their readme: https://github.com/HazyResearch/flash-attention

**Comment by drisspg** [comment]
Quick update: https://github.com/pytorch/pytorch/pull/100583 is all green. I will be adding dropout support and generic attention bias in follow up once this PR has landed

**Comment by usaxena-asapp** [comment]
@drisspg: I wanted to follow up on this:

> Correct this discussion is only around mem efficient attention. Currently FlashAttention does not support attention bias but you can see that it is on their roadmap posted on their readme: https://github.com/HazyResearch/flash-attention

This commit: https://github.com/HazyResearch/flash-attention/commit/40a25c8ee7465cf547b929cfa2937034e37bfce9 means that they're not going to support custom masks/arbitrary attention bias. 

Does that change anything for pytorch?

**Comment by drisspg** [comment]
Support for arbitrary attention bias is coming from the xformers memory-efficient attention kernel and not from the flash attention implementation

**Comment by ani300** [comment]
@drisspg After landing #100583, can we already use SDPA with biases and get the fast path, or is there something else that needs to land?

**Comment by drisspg** [comment]
One more PR is needed and that should be up imminetly

**Comment by StrangeTcy** [comment]
Please make a note or a comment or some other indication that the PR was merged?

**Comment by drisspg** [comment]
The PR has not been merged

**Comment by StrangeTcy** [comment]
Apologies, I might have phrased it wrongly; please make some kind of note here once it is?

**Comment by thegodone** [comment]
Do you have plan to add flash_attention 2 ?

**Comment by cpuhrsch** [comment]
@thegodone - Please see https://github.com/pytorch/pytorch/issues/105474

**Comment by drisspg** [comment]
I am going to close this because #104310 has been landed. This should allow users to create an alibi attn_mask and pass into sdpa. The mem efficient kernel will be called, which in my testing was ~ 2x faster than the math implementation

**Comment by ericzhou571** [comment]
I've tested out the new function, and it's impressive! When used during inference, it's significantly faster than the standard alibi attention implementation. However, I encountered a problem when integrating it into the training phase. Specifically, my loss value exploded, even though my training code was previously stable before I swapped the attention mechanism with this new version.

Here's the snippet of my code, how I use the new implementation:

```python
with torch.backends.cuda.sdp_kernel(enable_math=False):
    attn_output = F.scaled_dot_product_attention(query=query_states, key=key_states, value=value_states, attn_mask=attention_mask)
```

Is it possible that the new implementation currently only supports the forward pass and doesn't yet support the backward propagation with alibi?

**Comment by drisspg** [comment]
[ericzhou571](https://github.com/ericzhou571) So the numerics for the fused kernel are slightly different than that of the unfused math path. It is hard to say how much of a deviation can impact a model's training a priori, could you send some example inputs, sizes for the sdpa components. Otherwise it might be helpful to investigate, any local loss spikes that exist with the fused kernel and don't exist with unfused kernel

**Comment by Ken1256** [comment]
Does fast path now support all types of attention masks?

e.g: swin transformer
https://github.com/cszn/SCUNet
https://github.com/ofsoundof/GRL-Image-Restoration

**Comment by thomasahle** [comment]
@netw0rkf10w
This is not how you add ALiBi:
```
attn_weight = torch.softmax((Q @ K.transpose(-2, -1) * scale_factor) + attn_mask, dim=-1)
```
For ALiBi you need to add the attn_mask _inside_ the softmax, like this:
```
attn_weight = torch.softmax((Q @ K.transpose(-2, -1) * scale_factor + attn_mask), dim=-1)
```
(See page 5 in https://arxiv.org/pdf/2108.12409.pdf, or Figure 3).
I'm actually not sure if the attn_mask is multiplied by scale_factor or not. The paper is not clear on that. But I assume it's the way I wrote above.

I don't think there's a way to "simulate" adding attn_mask inside the softmax when adding it outside. If it was an exponential (instead of softmax) we could pull it out (though it would become multiplicative). With softmax I think we'd have to support an "inner" mask.

**Comment by vadimkantorov** [comment]
also, it seems that recently flash_attention started to natively support ALiBi via the `alibi_slopes` argument
